src/preprocess/links.rs

```diff
@@ -5,9 +5,35 @@ use utils::fs::file_to_string;
 use utils::take_lines;
 use errors::*;
 
+use super::Preprocessor;
+use book::{Book, BookItem};
+
 const ESCAPE_CHAR: char = '\\';
 
-pub fn replace_all<P: AsRef<Path>>(s: &str, path: P) -> Result<String> {
+pub struct ReplaceAllPreprocessor {
+    pub src_dir: PathBuf
+}
+
+impl Preprocessor for ReplaceAllPreprocessor {
+    fn run(&self, book: &mut Book) -> Result<()> {
+        for section in &mut book.sections {
+            match *section {
+                BookItem::Chapter(ref mut ch) => {
+                    let content = ::std::mem::replace(&mut ch.content, String::new());
```

Comment:
I don't think you need to use `mem::replace()` here. You could store the replaced text in a temporary (`let replaced = replace_all(&ch.content, base)`) then update `ch.content` on the next line.

Reply:
Good point, thanks